### PR TITLE
[RLlib] Issue 42396: Fix Algorithm.training_step() default implementation (multi-agent ALL_MODULES key causing error).

### LINKED ISSUE #42396: [RLlib] Algorithm.training_step fails after episode termination because `"__all__"` is considered a module name
### What happened + What you expected to happen

I'm trying to get a very basic RL Module working based on the examples and tests in the repo and I hit this issue.

When you run the attached reproduction script, it fails with:
```
Traceback (most recent call last):
  File "/Users/garymm/src/garymm/rl/./rl/ray/rllib_modules.py", line 139, in <module>
    train()
  File "/Users/garymm/src/garymm/rl/./rl/ray/rllib_modules.py", line 127, in train
    result = algo.train()
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/tune/trainable/trainable.py", line 342, in train
    raise skipped from exception_cause(skipped)
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/tune/trainable/trainable.py", line 339, in train
    result = self.step()
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/algorithms/algorithm.py", line 852, in step
    results, train_iter_ctx = self._run_one_training_iteration()
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/algorithms/algorithm.py", line 3042, in _run_one_training_iteration
    results = self.training_step()
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/algorithms/algorithm.py", line 1657, in training_step
    self.workers.sync_weights(
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/evaluation/worker_set.py", line 409, in sync_weights
    weights = weights_src.get_weights(policies)
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/core/learner/learner_group.py", line 466, in get_weights
    state = self._learner.get_module_state(module_ids)
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/core/learner/learner.py", line 806, in get_module_state
    module_states = self.module.get_state(module_ids)
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/core/rl_module/marl_module.py", line 280, in get_state
    return {
  File "/Users/garymm/mambaforge/envs/rl/lib/python3.10/site-packages/ray/rllib/core/rl_module/marl_module.py", line 281, in <dictcomp>
    module_id: self._rl_modules[module_id].get_state()
KeyError: '__all__'
```

I'm not sure if this is the right fix, but if I change this in Algorithm.training_step:

```py
self.workers.sync_weights(
    from_worker_or_learner_group=from_worker_or_trainer,
    policies=list(train_results.keys()),
    global_vars=global_vars,
)
```

to:

```py
self.workers.sync_weights(
    from_worker_or_learner_group=from_worker_or_trainer,
    policies=[p for p in train_results if p != "__all__"],
    global_vars=global_vars,
)
```

It seems to fix the problem. Not at all confident that's right now, I'm very new to this code base.

I'm not sure how to work around this without modifying Ray. I'll post if I figure out a work-around.

### Versions / Dependencies

2.9.0

### Reproduction script

```py
import argparse
import sys
from typing import Any, Mapping

import ray
import torch
from ray.rllib.algorithms import Algorithm, AlgorithmConfig
from ray.rllib.core.models.specs.typing import SpecType
from ray.rllib.core.learner.learner import LearnerHyperparameters
from ray.rllib.core.learner.torch.torch_learner import TorchLearner
from ray.rllib.core.rl_module.rl_module import (
    ModuleID,
    RLModule,
    RLModuleConfig,
    SingleAgentRLModuleSpec,
)
from ray.rllib.core.rl_module.torch.torch_rl_module import TorchRLModule
from ray.rllib.models.torch.torch_distributions import TorchCategorical
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.policy.torch_policy_v2 import TorchPolicyV2
from ray.rllib.utils.annotations import override
from ray.rllib.utils.nested_dict import NestedDict
from ray.rllib.utils.typing import TensorType

from rl.ray import log_filter


class MyModule(TorchRLModule):
    def __init__(self, config: RLModuleConfig) -> None:
        super().__init__(config)

    def setup(self):
        input_dim = self.config.observation_space.shape[0]
        hidden_dim = self.config.model_config_dict["fcnet_hiddens"][0]
        output_dim = self.config.action_space.n

        self.policy = torch.nn.Sequential(
            torch.nn.Linear(input_dim, hidden_dim),
            torch.nn.ReLU(),
            torch.nn.Linear(hidden_dim, output_dim),
        )

        self.input_dim = input_dim

    def get_train_action_dist_cls(self):
        return TorchCategorical

    def get_exploration_action_dist_cls(self):
        return TorchCategorical

    def get_inference_action_dist_cls(self):
        return TorchCategorical

    @override(RLModule)
    def output_specs_exploration(self) -> SpecType:
        return [SampleBatch.ACTION_DIST_INPUTS]

    @override(RLModule)
    def output_specs_inference(self) -> SpecType:
        return [SampleBatch.ACTION_DIST_INPUTS]

    @override(RLModule)
    def output_specs_train(self) -> SpecType:
        return [SampleBatch.ACTION_DIST_INPUTS]

    @override(RLModule)
    def _forward_inference(self, batch: NestedDict) -> Mapping[str, Any]:
        with torch.no_grad():
            return self._forward_train(batch)

    @override(RLModule)
    def _forward_exploration(self, batch: NestedDict) -> Mapping[str, Any]:
        with torch.no_grad():
            return self._forward_train(batch)

    @override(RLModule)
    def _forward_train(self, batch: NestedDict) -> Mapping[str, Any]:
        action_logits = self.policy(batch["obs"])
        return {SampleBatch.ACTION_DIST_INPUTS: action_logits}


class MyLearner(TorchLearner):
    def compute_loss_for_module(
        self,
        module_id: ModuleID,
        hps: LearnerHyperparameters,
        batch: NestedDict,
        fwd_out: Mapping[str, TensorType],
    ) -> Mapping[str, Any]:
        action_dist_inputs = fwd_out[SampleBatch.ACTION_DIST_INPUTS]
        action_dist_class = self._module[module_id].get_train_action_dist_cls()
        action_dist = action_dist_class.from_logits(action_dist_inputs)
        loss = -torch.mean(action_dist.logp(batch[SampleBatch.ACTIONS]))

        return loss


# Copied from ray/rllib/core/testing/bc_algorithm.py to remove TF dep
class MyAlgorithmConfig(AlgorithmConfig):
    def __init__(self, algo_class=None):
        super().__init__(algo_class=algo_class or MyAlgorithm)

    def get_default_rl_module_spec(self):
        return SingleAgentRLModuleSpec(module_class=MyModule)

    def get_default_learner_class(self):
        return MyLearner


class MyAlgorithm(Algorithm):
    @classmethod
    def get_default_policy_class(cls, config: AlgorithmConfig):
        return TorchPolicyV2


def train():
    algo_config: AlgorithmConfig = (
        MyAlgorithmConfig()
        .experimental(_enable_new_api_stack=True)
        .environment("CartPole-v1")
        .training(model={"fcnet_hiddens": [32, 32]})
    )

    algo: Algorithm = algo_config.build()
    _NUM_STEPS = 10
    for _ in range(_NUM_STEPS):
        result = algo.train()
    print("")
    print(result)


if __name__ == "__main__":
    sys.stderr = log_filter.filtered_stderr
    arg_parser = argparse.ArgumentParser()
    arg_parser.add_argument("--ray_addr", type=str)
    args = arg_parser.parse_args()
    ray.init(address=args.ray_addr)
    print(ray.available_resources())
    train()
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by sven1977**
Issue 42396: Fix Algorithm.training_step() default implementation (multi-agent ALL_MODULES key causing error).

## Why are these changes needed?

Closes #42396 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm.py [hunk 1]
```diff
@@ -1651,14 +1652,13 @@ def training_step(self) -> ResultDict:
...
-            # TODO (Kourosh): figure out how we are going to sync MARLModule
-            # weights to MARLModule weights under the policy_map objects?
             from_worker_or_trainer = None
             if self.config._enable_new_api_stack:
                 from_worker_or_trainer = self.learner_group
+
             self.workers.sync_weights(
                 from_worker_or_learner_group=from_worker_or_trainer,
-                policies=list(train_results.keys()),
+                policies=set(train_results.keys()) - {ALL_MODULES},
```
Comment (garymm) [inline rllib/algorithms/algorithm.py:1661]: Crying out for a comment at least.

**Review by garymm — COMMENTED** [review]
Thanks for the fix.
Would be nice for users if you tested this path too!

File: rllib/algorithms/algorithm.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm.py:1661]: Great catch @sven1977 !

**Review by simonsays1980 — APPROVED** [review]
LGTM

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**